app/questionnaire/path_finder.py

```diff
@@ -59,19 +59,32 @@ def build_path(self):
                 group_blocks = self._build_blocks_for_group(group, instance_idx)
```

Comment:
Might be worth grabbing the `group_instance` of the first block here. Then you wouldn't need to filter the `blocks` list on `group_id` outside the loop.

```
            first_group_instance_index = None
            for instance_idx in range(0, no_of_repeats):
                group_blocks = self._build_blocks_for_group(group, instance_idx)
                blocks += group_blocks

                if group_blocks and first_group_instance_index is None:
                    # get the group instance of the first instance of a block in this group that has not been skipped
                    first_group_instance_index = group_blocks[0]['group_instance']
```

Reply:
Doesnt work if first group_instance is skipped . Its not obvious though.

Reply:
But `group_blocks` doesn't contain skipped instances does it? So it should be safe to get the first one because that's the first one that wasn't skipped.

Reply:
I just changed it and the tests still pass.

Reply:
Oh that's just because `_build_blocks_for_group` is a generator.

That line would need to become

```
group_blocks = list(self._build_blocks_for_group(group, instance_idx))
```

Not going to labour the point now but I think it's clearer to do it here rather than extra it from the full block set later.